# Content List
### 01. Importing libraries
### 02. Importing data
### 03. Exercise Example Work
### 04. Task 4.7 Work

# 01. Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
# create path to main project folder
path = r'/Users/kirstinhelgeson/Desktop/Data Analytics Course/05-2023 Instacart Basket Analysis'

In [3]:
# import orders_products_merged.pkl file
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# 03. Exercise Example Work

In [4]:
# create a subset of df_ords_prods_merged that includes the first one million rows
df = df_ords_prods_merged[:1000000]

In [5]:
# define function for a 'price_label' column
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [6]:
# apply function to df subset
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/k6/pg6yg6t53cxfkd7vgpgz1q780000gn/T/ipykernel_4905/2545053247.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [7]:
# check values in new 'price_range' column
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [8]:
# check the maximum price within the df subset
df['prices'].max()

14.8

### Do the same process with loc() function instead

In [9]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/k6/pg6yg6t53cxfkd7vgpgz1q780000gn/T/ipykernel_4905/1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [10]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [11]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

### Do the same process on entire dataframe df_ords_prods_merged

In [13]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [14]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [15]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [16]:
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

In [17]:
# check column names in dataframe df_ords_prods_merged
print(df_ords_prods_merged.columns)

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merge', 'price_range_loc'],
      dtype='object')


In [18]:
# print the frequency of the 'orders_day_of_week' column. Note 0 = Saturday according to project brief.
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [19]:
# create for-loop to make a new column "busiest day"
result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [20]:
# print the result from above
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [21]:
# create a new column within df_ords_prods_merged dataframe and set equal to 'result'
df_ords_prods_merged['busiest_day'] = result

In [22]:
# check frequency of 'busiest day' column
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# 04. Task 4.7 Work

In [23]:
# Step 2. Create for-loop to make a new column "busiest days"
result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value in [0, 1]:
    result.append("Busiest days")
  elif value in [4, 3]:
    result.append("Least busy days")
  else:
    result.append("Regularly busy")

In [24]:
# print result from previous for-loop
result

['Regularly busy',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy days',
 'Regularly busy',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Least busy days',
 'Least busy days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Busiest day

In [25]:
# create a new column within df_ords_prods_merged dataframe and set equal to 'result'
df_ords_prods_merged['busiest_days'] = result

In [26]:
# Step 3. Check frequency of 'busiest days' column for accuracy
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy     12916111
Busiest days       11864412
Least busy days     7624336
Name: busiest_days, dtype: int64

### Step 3. The busiest days count of 11864412 is accurate according to the frequency of 'order_day_of_week' column performed previously. The least busy days count of 7624336 is also accurate. This makes it clear that the least busy days are significantly less busy than either the busiest days or the regularly busy days.

In [27]:
# print the frequency of 'order_hour_of_day' column. Note 0 = 12AM.
df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

### Client wants to identify busiest hours of the day with labels "Most orders", "Average orders", "Fewest orders". "Most orders" = [10, 11, 14, 15, 13, 12, 16, 9]. "Average orders" = [17, 8, 18, 19, 20, 7, 21, 22]. "Fewest orders" = [23, 6, 0, 1, 5, 2, 4, 3]

In [28]:
# Step 4. Create for-loop to make a new column "busiest_period_of_day"
result_2 = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_2.append("Most orders")
  elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
    result_2.append("Average orders")
  else:
    result_2.append("Fewest orders")

In [29]:
# print result_2 from previous for-loop
result_2

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [30]:
# Step 4. Create a new column within df_ords_prods_merged dataframe and set equal to 'result_2'
df_ords_prods_merged['busiest_period_of_day'] = result_2

In [31]:
# Step 5. Print frequency of 'busiest_period_of_day' column for accuracy
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

### 'Most orders' count of 21118071 matches up with frequency of 'order_hour_of_day' column printed above. 'Average orders' count of 9997651 also matches up. 'Fewest orders' count of 1289137 also matches up.

In [32]:
# print df_ords_prods_merged dataframe to see if any columns should be removed before exporting
df_ords_prods_merged

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                     NaN       42338   
32404857                  5                    15.0       42338   
32404858                 14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices _merge    price_range_loc  \
0               77              7     9.0   both  Mid-range product   
1               77              7     9.0   both  Mid-range product   
2               77              7     9.0   both  Mid-range product   
3               77              7     9.0   both  Mid-range product   
4               77              7     9.0   both  Mid-range product   
...            ...            ...     ...    ...                ...   
32404854        64              7     3.7   both  Low-range product   
32404855        64              7     3.7   both  Low-range product   
32404856        50             19     6.9   both  Mid-range product   
32404857        50             19     6.9   both  Mid-range product   
32404858       130             14    10.3   both  Mid-range product   

             busiest_day     busiest_days busiest_period_of_day  
0         Regularly busy   Regularly busy        Average orders  
1         Regularly busy  Least busy days        Average orders  
2         Regularly busy  Least busy days           Most orders  
3             Least busy  Least busy days        Average orders  
4             Least busy  Least busy days           Most orders  
...                  ...              ...                   ...  
32404854  Regularly busy   Regularly busy           Most orders  
32404855  Regularly busy   Regularly busy           Most orders  
32404856  Regularly busy   Regularly busy        Average orders  
32404857  Regularly busy  Least bu

### Since clients changed their minds about the labels created for "busiest_day" column, I will remove that column and keep the "busiest_days" column that has the updated labels per clients' request. I will also remove the "merge" column since it is not needed at this point after the original merge of the "ords" and "prods" dataframes.

In [33]:
# drop the 'busiest_day' and '_merge' columns from the df_ords_prods_merged dataframe
df_ords_prods_merged_2 = df_ords_prods_merged.drop(columns = ['busiest_day','_merge'])

In [34]:
# print df_ords_prods_merged_2 dataframe to verify correct columns were removed and all necessary information is still present before exporting
df_ords_prods_merged_2

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2398795        1             2                   3   
2           473747        1             3                   3   
3          2254736        1             4                   4   
4           431534        1             5                   4   
...            ...      ...           ...                 ...   
32404854   1320836   202557            17                   2   
32404855     31526   202557            18                   5   
32404856    758936   203436             1                   2   
32404857   2745165   203436             2                   3   
32404858   3093936   205420             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32404854                 15                     1.0       43553   
32404855                 11                     3.0       43553   
32404856                  7                     NaN       42338   
32404857                  5                    15.0       42338   
32404858                 14                     NaN       28818   

          add_to_cart_order  reordered                   product_name  \
0                         1          0                           Soda   
1                         1          1                           Soda   
2                         1          1                           Soda   
3                         1          1                           Soda   
4                         1          1                           Soda   
...                     ...        ...                            ...   
32404854                  2          1            Orange Energy Shots   
32404855                  2          1            Orange Energy Shots   
32404856                  4          0          Zucchini Chips, Pesto   
32404857                 16          1          Zucchini Chips, Pesto   
32404858                  8          0  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices    price_range_loc     busiest_days  \
0               77              7     9.0  Mid-range product   Regularly busy   
1               77              7     9.0  Mid-range product  Least busy days   
2               77              7     9.0  Mid-range product  Least busy days   
3               77              7     9.0  Mid-range product  Least busy days   
4               77              7     9.0  Mid-range product  Least busy days   
...            ...            ...     ...                ...              ...   
32404854        64              7     3.7  Low-range product   Regularly busy   
32404855        64              7     3.7  Low-range product   Regularly busy   
32404856        50             19     6.9  Mid-range product   Regularly busy   
32404857        50             19     6.9  Mid-range product  Least busy days   
32404858       130             14    10.3  Mid-range product  Least busy days   

         busiest_period_of_day  
0               Average orders  
1               Average orders  
2                  Most orders  
3               Average orders  
4                  Most orders  
...                        ...  
32404854           Most orders  
32404855           Most orders  
32404856        Average orders  
32404857         Fewest orders  
32404858           Most orders  

[32404859 rows x 16 columns]

In [35]:
# Step 7. export the df_ords_prods_merged_2 dataframe in pickle format
df_ords_prods_merged_2.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_new_variables.pkl'))